src/types.rs

```diff
@@ -1762,3 +1762,574 @@ impl std::fmt::Display for BasinName {
+            Operation::ReconfigureStream => Self::ReconfigureStream,
+            Operation::CheckTail => Self::CheckTail,
+            Operation::Append => Self::Append,
+            Operation::Read => Self::Read,
+            Operation::Trim => Self::Trim,
+            Operation::Fence => Self::Fence,
+        }
+    }
+}
+
+impl From<api::Operation> for Operation {
+    fn from(value: api::Operation) -> Self {
+        match value {
+            api::Operation::Unspecified => Self::Unspecified,
+            api::Operation::ListBasins => Self::ListBasins,
+            api::Operation::CreateBasin => Self::CreateBasin,
+            api::Operation::DeleteBasin => Self::DeleteBasin,
+            api::Operation::ReconfigureBasin => Self::ReconfigureBasin,
+            api::Operation::GetBasinConfig => Self::GetBasinConfig,
+            api::Operation::IssueAccessToken => Self::IssueAccessToken,
+            api::Operation::RevokeAccessToken => Self::RevokeAccessToken,
+            api::Operation::ListAccessTokens => Self::ListAccessTokens,
+            api::Operation::ListStreams => Self::ListStreams,
+            api::Operation::CreateStream => Self::CreateStream,
+            api::Operation::DeleteStream => Self::DeleteStream,
+            api::Operation::GetStreamConfig => Self::GetStreamConfig,
+            api::Operation::ReconfigureStream => Self::ReconfigureStream,
+            api::Operation::CheckTail => Self::CheckTail,
+            api::Operation::Append => Self::Append,
+            api::Operation::Read => Self::Read,
+            api::Operation::Trim => Self::Trim,
+            api::Operation::Fence => Self::Fence,
+        }
+    }
+}
+
+#[sync_docs]
+#[derive(Debug, Clone, Default)]
+pub struct AccessTokenScope {
+    pub basins: Option<ResourceSet>,
+    pub streams: Option<ResourceSet>,
+    pub tokens: Option<ResourceSet>,
+    pub op_groups: Option<PermittedOperationGroups>,
+    pub ops: HashSet<Operation>,
+}
+
+impl AccessTokenScope {
+    /// Create a new access token scope.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Overwrite resource set for tokens.
+    pub fn with_basins(self, basins: ResourceSet) -> Self {
+        Self {
+            basins: Some(basins),
+            ..self
+        }
+    }
+
+    /// Overwrite resource set for streams.
+    pub fn with_streams(self, streams: ResourceSet) -> Self {
+        Self {
+            streams: Some(streams),
+            ..self
+        }
+    }
+
+    /// Overwrite resource set for tokens.
+    pub fn with_tokens(self, tokens: ResourceSet) -> Self {
+        Self {
+            tokens: Some(tokens),
+            ..self
+        }
+    }
+
+    /// Overwrite operation groups.
+    pub fn with_op_groups(self, op_groups: PermittedOperationGroups) -> Self {
+        Self {
+            op_groups: Some(op_groups),
+            ..self
+        }
+    }
+
+    /// Overwrite operations.
+    pub fn with_ops(self, ops: Vec<Operation>) -> Self {
```

Comment:
Just a minor nit here:

```suggestion
    pub fn with_ops(self, ops: impl IntoIter<Item = Operation>) -> Self {
```